### ISSUE TITLE
Object Storage permissions using Bucket Policies

### ISSUE THREAD

**Issue by Kinuseka** [Chocobozzz/PeerTube#5497]
### Describe the problem to be solved

Object storage has been certainly a big part on some instances, and now on Peertube V5.0.0 allow to control our bucket's privacy so that it can no longer be access was certainly a big thing to prevent requests from someone that still knows the direct link. This was possible with the use of ACLs, it is an old feature by Amazon S3 which by them no longer recommends using and they actually encourage to use IAM and or policies instead. 

These are the object storage I have used in the past that I have come across

## Object Store With official support to ACLs
* Microsoft Azure Blob 
* Amazon Web Services, Simple Storage Services 
* Wasabi

## Object Store with limited support to ACLs
* Backblaze B2 (Cannot do ACLs object level)
* Cloudflare R2 (Cannot do ACL private)

## Object Store that denies support to ACLs
* Minio Object Storage

----
You can see that there are many inconsistencies with the API Support of ACLs  between various buckets, 
and even the cheapest one Backblaze B2 has incomplete support of it. While bigger companies with much expensive cost have a much complete API support, these definetly is a hurdle to the proper development of a proper video share especially those that use **Minio** with no support to ACLs whatsoever

### Describe the solution you would like

ACLs are definetly easier to work with but service providers often discourage them due to how "Expensive" they could be and could tangle the functionalities of a bucket.

## My Proposal
As a work around between these inconsistencies, we can do;

## **1. We can use bucket policies, as they are widely supported by any of those Object Storage providers I have mentioned above.**

These requires `PutBucketPolicy`  permissions

**Heres a working bucket policy that allows access to all files except `file.mp4`**
```json
{
    "Version": "2012-10-17",
    "Statement": [
        {
            "Effect": "Allow",
            "Principal": {
                "AWS": [
                    "*"
                ]
            },
            "Action": [
                "s3:GetObject"
            ],
            "Resource": [
                "arn:aws:s3:::test/*"
            ]
        },
        {
            "Effect": "Deny",
            "Principal": {
                "AWS": [
                    "*"
                ]
            },
            "Action": [
                "s3:GetObject"
            ],
            "Resource": [
                "arn:aws:s3:::test/file.mp4"
            ]
        }
    ]
}
```
Anything appended on the denied resources will be denied in anonymous access (aka public access without authentication)
while keeping everything else public. **I have tested this and it works as expected** 

Heres a flowchart of how I expect things to flow 
![image](https://user-images.githubusercontent.com/66762783/210130343-36650cdf-1505-452b-bb19-8ca86948f12e.png)

## **2. Create a new bucket containing the privated files.**

This solution may be resource intensive but is still a solution anyhow.
Public bucket will be exposed to the internet or via the CDN, and if the video is privated peertube will move it to a private bucket to store those privated video files. 

This will make any API calls to the Public bucket deny access as the file does not exist there any longer. This solution is easier to implement but might be resource intensive, so consider the bucket policy solution.

## **3. Append a random `hash` into the file name and or the directory of where the video is stored.**

assume the path for the video in bucket is `[BucketName]/streaming-playlists/hls/videoUUID/vidID-360-fragmented.mp4`
This path is a known public path to accessible to anyone. If a video turns **private** simply edit the directory name of the video by making it act as a **password** that only peertube server can know
and now it is:

`[BucketName]/streaming-playlists/hls/videoUUID[ADDARANDOMHASH]/vidID-360-fragmented.mp4`

of course the video is still public via the bucket but is now obscured and unknown.

**Comment by Kinuseka** [comment]
Started working on it myself, at [my own fork](https://github.com/Kinuseka/PeerTube_staging/tree/bucket-policy-exp) let me know if I should open a PR

**Comment by Chocobozzz** [comment]
Hello,

Policy size seems limited and I don't think that appending each file in the XML + reupload the policy will scale.

Do you know if all s3 providers support bucket policies based on object tags? (https://docs.aws.amazon.com/AmazonS3/latest/userguide/tagging-and-policies.html)

**Comment by Kinuseka** [comment]
Hi @kontrollanten 
continuing discussion from [here](https://github.com/Chocobozzz/PeerTube/pull/5502#issuecomment-1370186687)

Moving videos into a different folder requires copying and deleting them, this should be warned to users however that this could get expensive if they're using a Pay-as-you-go object storage. one example is with [wasabi's minimum storage ](https://wasabi-support.zendesk.com/hc/en-us/articles/360058734492-How-does-Wasabi-s-minimum-storage-duration-policy-work-)

**Comment by Kinuseka** [comment]
@dausruddin yes, Minio Does not support ACL headers, currently testing a very experimental fork [here](https://github.com/Kinuseka/PeerTube_staging/releases/tag/5.0.0UR-V2). But according to the S3 standards, Bucket policy is only limited to 20KiB which might not be scalable if you have thousands of users doing private uploads. For now you can do the work around by setting 

`object_storage.upload_acl.public = ""`
`object_storage.upload_acl.private = ""`

**Comment by Chocobozzz** [comment]
@dausruddin Please don't mix another issue with your problem. Let's continue the discussion on your origin issue: https://github.com/Chocobozzz/PeerTube/issues/5479

**Comment by stevespaw** [comment]
@dausruddin Why would you use 4 buckets? These would normally be 4 folders in the same bucket.
